src/HappyConnOpener.cc

```diff
@@ -869,3 +876,14 @@ HappyConnOpener::ranOutOfTimeOrAttempts() const
     return false;
 }
 
+void
+HappyConnOpener::Attempt::cancel(const char *reason)
+{
+    if (connector)
+        connector->cancel(reason);
+    if (connOpener.valid()) {
+        CallJobHere(17, 3, connOpener, Comm::ConnOpener, noteAbort);
+    }
```

Comment:
If having a non-nil `connector` implies that connOpener is set (but not necessarily valid), then I would simplify the above:

```suggestion
    if (connector) {
        connector->cancel(reason);
        CallJobHere(17, 3, connOpener, Comm::ConnOpener, noteAbort);
    }
```

IIRC, CallJobHere() will take care of pointers to gone jobs (i.e., it will work correctly when connOpener.valid() is false).